x/incentive/keeper/bond.go

```diff
@@ -7,94 +7,75 @@ import (
 ) {
 	bonded = coin.Zero(denom)
 	unbonding = coin.Zero(denom)
 	unbondings = []incentive.Unbonding{}
 
-	time := k.getLastRewardsTime(ctx)
+	time := k.GetLastRewardsTime(ctx)
+	duration := k.GetParams(ctx).UnbondingDuration
 
-	// sum all bonded and unbonding tokens for this denom, for the specified tier or across all tiers
-	for _, t := range bondTiers {
-		if tier == incentive.BondTierUnspecified || t == tier {
-			bonded = bonded.Add(k.getBonded(ctx, addr, denom, t))
+	// sum all bonded and unbonding tokens for this denom
+	bonded = bonded.Add(k.GetBonded(ctx, addr, denom))
 
-			tierUnbondings := k.getUnbondings(ctx, addr, denom, t)
-			for _, u := range tierUnbondings {
-				if u.End > time {
-					// this unbonding is still ongoing, and can be counted normally
-					unbondings = append(unbondings, u)
-					unbonding = unbonding.Add(u.Amount)
-				}
-				// Otherwise, this unbonding is completed, and will be quietly cleared by the next state-altering message
-				// which is permitted to affect this account. For now, it is omitted from the returned unbonding total.
-			}
+	storedUnbondings := k.getUnbondings(ctx, addr, denom)
+	for _, u := range storedUnbondings {
+		if u.End > time && u.Start+duration > time {
+			// If unbonding has passed neither its original end time nor its dynamic end time based on parameters
+			// the unbonding is still ongoing, and can be counted normally.
+			// This logic allows a reduction in unbonding duration param to speed up existing unbondings.
+			unbondings = append(unbondings, u)
+			unbonding = unbonding.Add(u.UToken)
 		}
+		// Otherwise, this unbonding is completed, and will be quietly cleared by the next state-altering message
+		// which is permitted to affect this account. For now, it is omitted from the returned unbonding total.
 	}
 
 	// returned values reflect the real situation (after completed unbondings), not what is currently stored in state
 	return bonded, unbonding, unbondings
 }
 
-// increaseBond increases the bonded uToken amount for an account at a given tier, and updates the module's total.
+// increaseBond increases the bonded uToken amount for an account, and updates the module's total.
 // it also initializes the account's reward tracker if the bonded amount was zero. This function should only be
 // called during MsgBond.
-func (k Keeper) increaseBond(ctx sdk.Context, addr sdk.AccAddress, tier incentive.BondTier, bond sdk.Coin) error {
+func (k Keeper) increaseBond(ctx sdk.Context, addr sdk.AccAddress, bond sdk.Coin) error {
 	// get bonded amount before adding the currently bonding tokens
-	bonded := k.getBonded(ctx, addr, bond.Denom, tier)
+	bonded := k.GetBonded(ctx, addr, bond.Denom)
 	// if bonded amount was zero, reward tracker must be initialized
 	if bonded.IsZero() {
-		if err := k.UpdateRewardTracker(ctx, addr, tier, bond.Denom); err != nil {
+		if err := k.updateRewardTracker(ctx, addr, bond.Denom); err != nil {
```

Comment:
shouldn't we always update the reward tracker?

Reply:
Usually it is updated when claiming rewards.

The edge case here is when creating a bond for the first time, tracker must be initialized.

Reply:
OK, then let's update the function comment that the reward tracker must be updated before this function is called. Also, for that edge case, can we move initialization to the caller, where that edge case occurs? Maybe during the creation?

Reply:
I put it in here so it's impossible to forget to initialize no matter when/how the function is called. It should be more bug-proof that way, in the long term.